File: RateLimiter.py
import threading
import time
from typing import Dict, Any


class RateLimiter:
    def __init__(self, rate: float, capacity: float):
        if rate <= 0 or capacity <= 0:
            raise ValueError("Rate and Capacity must both be positive")

        self.rate = rate
        self.capacity = capacity
        self.clients: Dict[str, Dict[str, Any]] = {}
        self.lock = threading.Lock()

    def _get_or_create_client_state(self, client_id, timestamp):
        if client_id not in self.clients:
            self.clients[client_id] = {
                "tokens": self.capacity,
                "last_updated": timestamp,
            }
        return self.clients[client_id]

    def is_allowed(self, client_id: str) -> bool:
        with self.lock:
            now = time.monotonic()
            client_state = self._get_or_create_client_state(client_id, now)
            elapsed = now - client_state["last_updated"]

            client_state["tokens"] = min(
                self.capacity, client_state["tokens"] + elapsed * self.rate
            )
            client_state["last_updated"] = now

            print(1.0 - 1e-9)
            if client_state["tokens"] >= 1.0 - 1e-9:
                client_state["tokens"] -= 1.0
                print(f"tokens = {self.clients[client_id]}")
                return True
            print(f"tokens = {self.clients[client_id]}")
            return False


rl = RateLimiter(1.0, 5.0)

test = rl.is_allowed("test")

File: RateLimiter.ts
interface ClientState {
	tokens: number;
	lastUpdated: number;
};
interface Tokens {
	clientId: ClientState;
};

class RateLimiter {
	private readonly rate: number;
	private readonly capacity: number;
	private tokens: Tokens;
	constructor(rate: number, capacity: number) {
		this.rate = rate;
		this.capacity = capacity;
		this.tokens = {}
	};

	private _getOrCreateClientState(clientId: string, time: number): ClientState {
		let value = this.tokens.get()
		if (this.tokens.clientId) {
			k
		}

		if (this.tokens[clientId])
			return clientState
	};

	isAllowed(clinet_id: string): boolean {
		// need to lock threads
		// get current time
		const now = Date.now();
		// get client state
		let clientState = this._getOrCreateClientState(clinet_id, now);
		// get elapesed time
		const elapsed = now - clientState.last_updated;
		// calculate tokens based on smallest of capacity or state.tokens + elapsed + rate
		// store updated token value
		//
		//if tokens available 
		//	reduce token count
		//	return true
		//
		//if tokens unavailable
		//	return false
		return false

	};

}
